File: placement-reactnative/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/data/PlacementViewData.kt
package com.storylyplacementreactnative.common.data

import com.appsamurai.storyly.banner.analytics.event.BannerEventPayload
import com.appsamurai.storyly.banner.data.model.BannerPayload
import com.appsamurai.storyly.core.analytics.error.STRErrorPayload
import com.appsamurai.storyly.core.analytics.event.STREventPayload
import com.appsamurai.storyly.core.data.model.STRPayload
import com.appsamurai.storyly.storybar.analytics.event.StoryBarEventPayload
import com.appsamurai.storyly.storybar.data.model.StoryBarPayload
import com.appsamurai.storyly.swipecard.analytics.event.SwipeCardEventPayload
import com.appsamurai.storyly.swipecard.data.model.SwipeCardPayload
import com.appsamurai.storyly.videofeed.analytics.event.VideoFeedEventPayload
import com.appsamurai.storyly.videofeed.data.model.VideoFeedPayload
import com.storylyplacementreactnative.common.data.widgets.encodeBannerPayload
import com.storylyplacementreactnative.common.data.widgets.encodeStoryBarPayload
import com.storylyplacementreactnative.common.data.widgets.encodeSwipeCardPayload
import com.storylyplacementreactnative.common.data.widgets.encodeVideoFeedPayload


fun encodeSTRPayload(payload: STRPayload): Map<String, Any?> {
    return when (payload) {
        is StoryBarPayload -> encodeStoryBarPayload(payload)
        is VideoFeedPayload -> encodeVideoFeedPayload(payload)
        is BannerPayload -> encodeBannerPayload(payload)
        is SwipeCardPayload -> encodeSwipeCardPayload(payload)
        else -> null
    } ?: emptyMap()
}

fun encodeSTREventPayload(payload: STREventPayload): Map<String, Any?> {
    val baseMap = mapOf(
        "event" to payload.baseEvent.getType()
    )
    val addition = when (payload) {
        is StoryBarEventPayload -> encodeStoryBarPayload(payload.payload)
        is VideoFeedEventPayload -> encodeVideoFeedPayload(payload.payload)
        is BannerEventPayload -> encodeBannerPayload(payload.payload)
        is SwipeCardEventPayload -> encodeSwipeCardPayload(payload.payload)
        else -> null
    } ?: emptyMap()
    return baseMap + addition
}

fun encodeSTRErrorPayload(payload: STRErrorPayload): Map<String, Any?> {
    return mapOf(
        "event" to payload.baseError.getType(),
        "message" to payload.message,
    )
}

File: placement-reactnative/storyly-placement-react-native/ios/Common/Data/PlacementViewData.swift
import Foundation
import StorylyPlacement
import StorylyCore
import StorylyBanner
import StorylyStoryBar
import StorylyVideoFeed
import StorylySwipeCard
import StorylyCanvas


// MARK: - STRPayload Encoding

func encodeSTRPayload(_ payload: STRPayload) -> [String: Any] {
    switch payload {
    case let storyBarPayload as StoryBarPayload:
        return encodeStoryBarPayload(storyBarPayload) ?? [:]
    case let videoFeedPayload as VideoFeedPayload:
        return encodeVideoFeedPayload(videoFeedPayload) ?? [:]
    case let bannerPayload as BannerPayload:
        return encodeBannerPayload(bannerPayload) ?? [:]
    case let swipeCardPayload as SwipeCardPayload:
        return encodeSwipeCardPayload(swipeCardPayload) ?? [:]
    case let canvasPayload as CanvasPayload:
        return encodeCanvasPayload(canvasPayload) ?? [:]
    default:
        return [:]
    }
}

// MARK: - STREventPayload Encoding

func encodeSTREventPayload(_ payload: STREventPayload) -> [String: Any] {
    let baseMap: [String: Any] = [
        "event": payload.baseEvent.getType()
    ]
    
    var addition: [String: Any] = [:]
    switch payload {
    case let storyBarEvent as StoryBarEventPayload:
        addition = encodeStoryBarPayload(storyBarEvent.payload) ?? [:]
    case let videoFeedEvent as VideoFeedEventPayload:
        addition = encodeVideoFeedPayload(videoFeedEvent.payload) ?? [:]
    case let bannerEvent as BannerEventPayload:
        addition = encodeBannerPayload(bannerEvent.payload) ?? [:]
    case let swipeCardEvent as SwipeCardEventPayload:
        addition = encodeSwipeCardPayload(swipeCardEvent.payload) ?? [:]
    case let canvasEvent as CanvasEventPayload:
        addition = encodeCanvasPayload(canvasEvent.payload) ?? [:]
    default:
        break
    }
    
    return baseMap.merging(addition) { (_, new) in new }
}

// MARK: - STRErrorPayload Encoding

func encodeSTRErrorPayload(_ payload: STRErrorPayload) -> [String: Any] {
    return [
        "event": payload.baseError.getType(),
        "message": payload.message
    ]
}

// MARK: - Widget Controller Encoding

func encodeWidgetController(_ controller: STRWidgetController, widgetMap: inout [String: WeakReference<STRWidgetController>]) -> [String: String] {
    return [
        "type": controller.getType().description,
        "viewId": updateWidgetMapKey(controller, widgetMap: &widgetMap)
    ]
}

func updateWidgetMapKey(_ controller: STRWidgetController, widgetMap: inout [String: WeakReference<STRWidgetController>]) -> String {
    widgetMap = widgetMap.filter { $0.value.value != nil }
    
    if let existingKey = widgetMap.first(where: { $0.value.value === controller })?.key {
        return existingKey
    }
    
    let newKey = UUID().uuidString
    widgetMap[newKey] = WeakReference(value: controller)
    return newKey
}

// MARK: - WeakReference Helper

class WeakReference<T: AnyObject> {
    weak var value: T?
    
    init(value: T) {
        self.value = value
    }
}

